src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/DependencyIconSetCache.cs

```diff
@@ -52,6 +52,7 @@ private static T GetOrAdd<T>(ref ImmutableHashSet<T> location, T value)
                 }
 
                 // Optimistic update failed, so loop around and try again.
+                prior = original;
```

Comment:
Why can't this entire method be replaced with ApplyChangeOptimistically?

Reply:
The improvement here is lookup within the loop.

Reply:
`ImmutableInterlocked` has `GetOrAdd` methods for `ImmutableDictionary` but not `ImmutableHashSet`. At some point I'll file a PR with this addition for corefx.

Reply:
Hmm, but I like tried and tested methods also, avoids bugs. 

Did this actually show up on the radar? If so we should be pushing these changes into VS threading.

Reply:
I tried pushing a similar change into vs-threading. The response suggested this kind of thing belongs in `ImmutableInterlocked`, and I tend to agree.